internal/controller/object/object.go

```diff
@@ -247,50 +279,73 @@ func (c *connector) Connect(ctx context.Context, mg resource.Managed) (managed.E
 		sanitizeSecrets: c.sanitizeSecrets,
 
 		kindObserver: c.kindObserver,
-	}, nil
+		syncer: &PatchingResourceSyncer{
+			client: resource.ClientApplicator{
+				Client:     k,
+				Applicator: resource.NewAPIPatchingApplicator(k),
+			},
+		},
+	}
+
+	if c.ssaEnabled {
+		dc, err := discovery.NewDiscoveryClientForConfig(rc)
```

Comment:
This looks wrong. It is called on every reconcile, isn't it? Instantiating a new discovery client is heavy.

Reply:
The problem is, this is not something static, we need to initialize discovery client for the kubeconfig configured in the ProviderConfig referenced in the Object. It can change anytime, so, we need to initialize during reconcile.

Reply:
We have been digging a bit deeper. The extractor below downloads the OpenAPI v2 schema via `dc`. This is a big (multi-megabyte) proto file. It will grow even more (a lot) if you have many CRDs from providers installed. And the reconciler will do that on every reconcile, i.e. tens-of-megabyte download and unmarshalling. This is likely a blocker for the approach 🔴.

The solution might be caching of the discovery client by provider config.

Reply:
https://github.com/turkenh/provider-kubernetes/pull/3 is merged into this feature branch, that implements a custom CachingUnstructuredExtractor, which utilizes OpenAPI v3 discovery and caching per GV.

Reply:
The original blocker problem is solved with the caching.